server/sonar-server/src/main/java/org/sonar/server/computation/ComputationStepExecutor.java

```diff
@@ -19,70 +19,80 @@
  */
 package org.sonar.server.computation;
 
-import com.google.common.base.Throwables;
-import org.sonar.api.server.ServerSide;
-import org.sonar.api.utils.System2;
 import org.sonar.api.utils.log.Logger;
-import org.sonar.api.utils.log.Loggers;
 import org.sonar.core.util.logs.Profiler;
-import org.sonar.server.computation.activity.ActivityManager;
 import org.sonar.server.computation.monitoring.CEQueueStatus;
 import org.sonar.server.computation.step.ComputationStep;
-import org.sonar.server.computation.step.ComputationSteps;
 
 import static java.lang.String.format;
-import static org.sonar.db.compute.AnalysisReportDto.Status.FAILED;
-import static org.sonar.db.compute.AnalysisReportDto.Status.SUCCESS;
+import static java.util.Objects.requireNonNull;
 
-@ServerSide
-public class ComputationService {
-
-  private static final Logger LOG = Loggers.get(ComputationService.class);
-
-  private final ReportQueue.Item item;
-  private final ComputationSteps steps;
-  private final ActivityManager activityManager;
-  private final System2 system;
+public final class ComputationStepExecutor {
+  private final Logger logger;
+  private final Listener listener;
+  private final String description;
   private final CEQueueStatus queueStatus;
 
-  public ComputationService(ReportQueue.Item item, ComputationSteps steps, ActivityManager activityManager, System2 system,
-    CEQueueStatus queueStatus) {
-    this.item = item;
-    this.steps = steps;
-    this.activityManager = activityManager;
-    this.system = system;
+  public ComputationStepExecutor(Logger logger, Listener listener, String description, CEQueueStatus queueStatus) {
+    this.logger = requireNonNull(logger);
+    this.listener = requireNonNull(listener);
+    this.description = requireNonNull(description);
     this.queueStatus = queueStatus;
   }
 
-  public void process() {
+  public void execute(Iterable<ComputationStep> steps) {
     queueStatus.addInProgress();
-    String projectKey = item.dto.getProjectKey();
-    String message = format("Analysis of project %s (report %d)", projectKey, item.dto.getId());
-    Profiler profiler = Profiler.create(LOG).startDebug(message);
-
+    listener.onStart();
+    Profiler profiler = Profiler.create(logger).startDebug(description);
     long timingSum = 0L;
-    Profiler stepProfiler = Profiler.create(LOG);
+    Profiler stepProfiler = Profiler.create(logger);
     try {
-      for (ComputationStep step : steps.instances()) {
+      for (ComputationStep step : steps) {
         stepProfiler.start();
         step.execute();
         timingSum += stepProfiler.stopInfo(step.getDescription());
       }
-      item.dto.setStatus(SUCCESS);
-      long timing = logProcessingEnd(message, profiler, timingSum);
+      long timing = logProcessingEnd(description, profiler, timingSum);
       queueStatus.addSuccess(timing);
+      listener.onSuccess(timing);
     } catch (Throwable e) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Catch Exception instead of Throwable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1181)